chain/pool/txwithtime.go

```diff
@@ -0,0 +1,229 @@
+package pool
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/nknorg/nkn/v2/common"
+	"github.com/nknorg/nkn/v2/pb"
+	"github.com/nknorg/nkn/v2/transaction"
+)
+
+// sync txn
+const MaxSyncTxnInterval = 16000 // in milli seconds
+
+type txnWithTime struct { // abbreviation is twt: txn with time
+	arriveTime int64
+	txn        *transaction.Transaction
+}
+
+// xor of byte slices.
+// notice!!!, b1 will be modified, b1 is the result.
+func xorBytes(b1 []byte, b2 []byte) ([]byte, error) {
+	if len(b1) != len(b2) {
+		return nil, fmt.Errorf("Input slices don't have the same length")
+	}
+
+	for i := 0; i < len(b1); i++ {
+		b1[i] = b1[i] ^ b2[i]
+	}
+
+	return b1, nil
+}
+
+// update twtFingerprint
+func (tp *TxnPool) updateTwtFingerprint(txnHash common.Uint256) []byte {
+
+	if tp.twtFingerprint == nil {
+		tp.twtFingerprint = make([]byte, 0, len(txnHash.ToArray()))
+		tp.twtFingerprint = append(tp.twtFingerprint, txnHash.ToArray()...)
+	} else {
+		tp.twtFingerprint, _ = xorBytes(tp.twtFingerprint, txnHash.ToArray())
+	}
+	return tp.twtFingerprint
+}
+
+// when append new txn to pool, we add this txn pointer to txtWithTime map too.
+func (tp *TxnPool) appendTwt(txn *transaction.Transaction) *txnWithTime {
+
+	now := time.Now().UnixMilli()
+
+	twt := &txnWithTime{
+		arriveTime: now,
+		txn:        txn,
+	}
+
+	tp.twtMapMu.Lock()
+	defer tp.twtMapMu.Unlock()
+
+	tp.twtMap.Set(txn.Hash(), twt)
```

Comment:
Check return value, only update hash and count if no current value

Reply:
Actually we should get first, and only call set when key is not in map. This way we can guarantee insert order is strictly the same as arrive time.

Reply:
Yes, modified.